// kadira-core/dist/clock.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _isomorphicFetch = require('isomorphic-fetch');

var _isomorphicFetch2 = _interopRequireDefault(_isomorphicFetch);

var _retry = require('./retry.js');

var _retry2 = _interopRequireDefault(_retry);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEFAULTS = {
  endpoint: ''
};

var Clock = function () {
  function Clock(_options) {
    (0, _classCallCheck3.default)(this, Clock);

    this._options = (0, _assign2.default)({}, DEFAULTS, _options);
    this._diff = 0;
    this.ready = false;
  }

  (0, _createClass3.default)(Clock, [{
    key: 'getTime',
    value: function getTime() {
      // current time on Kadira server
      return this._clientTS() + this._diff;
    }
  }, {
    key: 'fixTime',
    value: function fixTime(timestamp) {
      // `timestamp` on Kadira server
      return timestamp + this._diff;
    }
  }, {
    key: 'sync',
    value: function sync() {
      var _this = this;

      // calculate the time difference
      return (0, _retry2.default)(function () {
        return _this._syncOnce();
      });
    }
  }, {
    key: '_clientTS',
    value: function _clientTS() {
      // Get client timestamp while considering time related
      // libraries messing up the Date object. "Y U DO THIS?"
      var now = Date.now();
      if (typeof now === 'number') {
        return now;
      }

      // some time related libraries screw up Date.now()
      // and it returns a Date object instead of a number
      if (now instanceof Date) {
        return now.getTime();
      }

      // final attempt to get time
      return new Date().getTime();
    }
  }, {
    key: '_syncOnce',
    value: function _syncOnce() {
      var _this2 = this;

      var startTS = undefined;

      return this._fetchTime().then(function () {
        startTS = _this2._clientTS();
        return _this2._fetchTime();
      }).then(function (serverTS) {
        var latency = (_this2._clientTS() - startTS) / 2;
        _this2._diff = serverTS - latency - startTS;
        _this2.ready = true;
      });
    }
  }, {
    key: '_fetchTime',
    value: function _fetchTime() {
      return (0, _isomorphicFetch2.default)(this._options.endpoint).then(function (res) {
        if (res.status !== 200) {
          throw new Error('request failed: ' + res.status);
        }

        return res.text().then(function (txt) {
          return parseInt(txt, 10);
        });
      });
    }
  }]);
  return Clock;
}();

exports.default = Clock;
